### PR TITLE
Remember the eval mode of submodules when switching trainer stages

### LINKED ISSUE #18930: Identify parts of the models that are set in '.eval()' mode during initialization and keep them in eval mode after validation epoch ends
### Description & Motivation

Currently, if I want a submodule of my model completely **frozen** during training, I cannot do so because even with 

1. Setting this submodule in `.eval()` mode in `__init__()`
2. Setting all `required_grad` of all parameters in this submodule to `False`
3. Using `torch.no_grad()` context when passing data through this module

If my submodule contains ANY batch-normalization layers, these will not stay in `.eval()` mode because the whole model (which contains this submodule) is set to `.train()` mode after validation epoch ends. Hence, these batch-normalization layers' running stats will change and affect the performance and training of the entire model.

This behavior is very easy to create a minimal example to see this effect.

### Pitch

Search for params that are set in eval mode during initialization and keep them in eval mode after validation epoch ends.

### Alternatives

_No response_

### Additional context

_No response_

cc @borda

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #18930
Part of #16827

A common issue users are facing is that the loop calls `train()` on the LightningModule despite the user having frozen certain layers. For example,
- user finetunes a model and freezes certain layers they don't want to train
- user's model has a feature extractor they don't train

This leads to a surprise when the user finds out that their batch norm layers have changed statistics, even though they were set explicitly to `eval()` mode. To avoid this, the user has to learn that they should override the `on_validation_model_eval()` and `on_validation_model_train()` hooks in the module, but this is a detail difficult to find in our docs and get right. Most users who face this challenge end up on slack or GH to ask for help.

The PR makes the following changes to automate this for the user:
- The validation loop captures the `.training` mode of every submodule before calling `.eval()` now. When the validation loop ends, and before switching to training, it restores the `.training` mode on all submodules to what it was before. This ensures that layers the user has chosen to be in eval mode **remain** in eval mode!
- The fit-loop no longer calls `.train()` at the beginning with the same motivation: The user can now set a subset of their model to `.eval()` mode / freeze it explicitly in the LightningModule's `__init__` without doing acrobatics with hooks, and the Trainer will respect it and preserve it (see the added test). Note: This is not a breaking change, because PyTorch's default is to have a model in `.training=True` mode.

----
:books: Documentation preview :books:: https://pytorch-lightning--18951.org.readthedocs.build/en/18951/

cc @borda @justusschock @awaelchli

### CODE REVIEW

File: tests/tests_pytorch/models/test_hooks.py [hunk 1]
```diff
@@ -470,18 +470,14 @@ def training_step(self, batch, batch_idx):
         {"name": "Callback.on_sanity_check_start", "args": (trainer, model)},
         {"name": "val_dataloader"},
         {"name": "train", "args": (False,)},
-        {"name": "on_validation_model_eval"},
```
Comment (awaelchli) [inline tests/tests_pytorch/models/test_hooks.py:473]: Now that these hooks are only called when explicitly overridden by the user, I should extend/parameterize the test to make this conditional. @carmocca Do you have some idea how I could conditionally define an overridden hook on the HookedModel defined here?

File: src/lightning/pytorch/utilities/model_helpers.py [hunk 2]
```diff
@@ -58,6 +58,24 @@ def get_torchvision_model(model_name: str, **kwargs: Any) -> nn.Module:
...
+
+    def capture(self, module: nn.Module) -> None:
+        for name, mod in module.named_modules():
+            self.mode[name] = mod.training
+
+    def restore(self, module: nn.Module) -> None:
+        for name, mod in module.named_modules():
+            if name not in self.mode:
+                # TODO: warning cache?
+                continue
```
Comment (awaelchli) [inline src/lightning/pytorch/utilities/model_helpers.py:75]: This case can only occur if the model changed it's composition between `capture()` and `restore()`. How should we handle this?

- Ignore (current)
- A warning
- other?

File: src/lightning/pytorch/utilities/model_summary/model_summary.py [hunk 3]
```diff
@@ -300,7 +302,7 @@ def _forward_example_input(self) -> None:
                 model(**input_)
             else:
                 model(input_)
-        model.train(mode)  # restore mode of module
+        mode.restore(model)
```
Comment (awaelchli) [inline src/lightning/pytorch/utilities/model_summary/model_summary.py:305]: TODO: Extend model summary test

File: src/lightning/pytorch/utilities/model_helpers.py [hunk 4]
```diff
@@ -58,6 +58,24 @@ def get_torchvision_model(model_name: str, **kwargs: Any) -> nn.Module:
...
 
 
+class _ModuleMode:
+    """Captures the ``nn.Module.training`` (bool) mode of every submodule, and allows it to be restored later on."""
+
+    def __init__(self) -> None:
+        self.mode: Dict[str, bool] = {}
+
+    def capture(self, module: nn.Module) -> None:
+        for name, mod in module.named_modules():
```
Comment (carmocca) [inline src/lightning/pytorch/utilities/model_helpers.py:68]: Should the dict be cleared before capturing in case the model changes?

File: src/lightning/pytorch/utilities/model_helpers.py (hunk 2 above)
Comment (carmocca) [reply src/lightning/pytorch/utilities/model_helpers.py:75]: I vote for ignoring. It could be a debug-level message

File: tests/tests_pytorch/models/test_hooks.py (hunk 1 above)
Comment (carmocca) [reply tests/tests_pytorch/models/test_hooks.py:473]: You can simply override it on the class ([example](https://github.com/Lightning-AI/lightning/blob/feature/eval-mode/tests/tests_pytorch/models/test_hooks.py#L393-L395)) or monkeypatch it on the class if you want to do it based on a pytest parametrization value.

File: src/lightning/pytorch/loops/evaluation_loop.py [hunk 5]
```diff
@@ -293,13 +294,21 @@ def _on_evaluation_model_eval(self) -> None:
         """Sets model to eval mode."""
         trainer = self.trainer
         hook_name = "on_test_model_eval" if trainer.testing else "on_validation_model_eval"
-        call._call_lightning_module_hook(trainer, hook_name)
+        if is_overridden(hook_name, trainer.lightning_module):
+            call._call_lightning_module_hook(trainer, hook_name)
+        else:
+            self._module_mode.capture(trainer.lightning_module)
+            assert trainer.model is not None
+            trainer.model.eval()
```
Comment (carmocca) [inline src/lightning/pytorch/loops/evaluation_loop.py:302]: Why don't we still call the hook unconditionally but capture the state before?

```suggestion
        self._module_mode.capture(trainer.lightning_module)
        call._call_lightning_module_hook(trainer, hook_name)
```

**Review by carmocca — COMMENTED** [review]
Does this PR replace https://github.com/Lightning-AI/lightning/pull/18826?

File: src/lightning/pytorch/utilities/model_helpers.py (hunk 4 above)
Comment (awaelchli) [reply src/lightning/pytorch/utilities/model_helpers.py:72]: Good idea! Done and added test.

**Comment by awaelchli** [comment]
> Does this PR replace https://github.com/Lightning-AI/lightning/pull/18826?

@carmocca Good find! Yes in fact it does, I verified thanks to the repro example that was posted there. After this PR lands, we could still use https://github.com/Lightning-AI/lightning/pull/18826 to add an additional test case for the tuner.

File: tests/tests_pytorch/models/test_hooks.py (hunk 1 above)
Comment (awaelchli) [reply tests/tests_pytorch/models/test_hooks.py]: I decided to override it by default, and then conditionally set the `model.hook = None` for the exclusion.

**Review by carmocca — APPROVED** [review]

**Comment by adosar** [comment]
@awaelchli

> This leads to a surprise when the user finds out that their batch norm layers have changed statistics, even though they were set explicitly to eval() mode. To avoid this, the user has to learn that they should override the on_validation_model_eval() and on_validation_model_train()

Since the problem was in the training loop which called `train()`, even if the user sets them in `on_validation_model_{eval|train}`, it would lead to incorrect behavior (before this PR). Am I right?  (just asking for better understanding `Lightning`)

Also, currently in the [docs](https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#validation-loop) it is shown that `model.train()` is called after the validation ends what was done in this PR.

```python
# ...
for batch_idx, batch in enumerate(train_dataloader):
    loss = model.training_step(batch, batch_idx)
    loss.backward()
    # ...

    if validate_at_some_point:
        # disable grads + batchnorm + dropout
        torch.set_grad_enabled(False)
        model.eval()

        # ----------------- VAL LOOP ---------------
        for val_batch_idx, val_batch in enumerate(val_dataloader):
            val_out = model.validation_step(val_batch, val_batch_idx)
        # ----------------- VAL LOOP ---------------

        # enable grads + batchnorm + dropout
        torch.set_grad_enabled(True)
        model.train()
```

I would happy to open a new issue for fixing the docs to better align with this nice PR!
